tests/DCPS/Compiler/xcdr/xcdr.cpp

```diff
@@ -381,6 +433,97 @@ TEST(basic_tests, MutableXcdr12Struct)
   baseline_checks<MutableXcdr12Struct>(xcdr2, mutable_xcdr2_struct_expected);
 }
 
+// Union Tests =============================================================
+
+const unsigned char mutable_xcdr2_union_expected_short[] = {
+  // Delimiter
+  0x00, 0x00, 0x00, 0x0e, // +4 = 4
+  // Discriminator EMHEADER
+  0x20, 0x00, 0x00, 0x00, // PID  +4 = 8
+  // Discriminator
+  0x00, 0x00, 0x00, 0x00, // +4 = 12
+  // short_field
+  0x10, 0x00, 0x00, 0x00, // PID  +4 = 16
+  0x7f, 0xff // +2 = 18
+};
+
+const unsigned char mutable_xcdr2_union_expected_long[] = {
+  // Delimiter
+  0x00, 0x00, 0x00, 0x10, // +4 = 4
+  // Discriminator EMHEADER
+  0x20, 0x00, 0x00, 0x00, // PID  +4 = 8
+  // Discriminator
+  0x00, 0x00, 0x00, 0x01, // +4 = 12
+  // long_field
+  0x20, 0x00, 0x00, 0x01,// PID  +4 = 16
+  0x7f, 0xff, 0xff, 0xff // +4 = 20
+};
+
+const unsigned char mutable_xcdr2_union_expected_octet[] = {
+  // Delimiter
+  0x00, 0x00, 0x00, 0x0d, // +4 = 4
+  // Discriminator EMHEADER
+  0x20, 0x00, 0x00, 0x00, // PID  +4 = 8
+  // Discriminator
+  0x00, 0x00, 0x00, 0x02, // +4 = 12
+  // octet_field
+  0x00, 0x00, 0x00, 0x02, // PID  +4 = 16
+  0x01                    // +1 = 17
+};
+
+const unsigned char mutable_xcdr2_union_expected_long_long[] = {
+  // Delimiter
+  0x00, 0x00, 0x00, 0x14, // +4 = 4
+  // Discriminator EMHEADER
+  0x20, 0x00, 0x00, 0x00, // PID  +4 = 8
+  // Discriminator
+  0x00, 0x00, 0x00, 0x03, // +4 = 12
+  // long_field
+  0x30, 0x00, 0x00, 0x03, // PID  +4 = 16
+  0x7f, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff  // +8 = 24
+};
+
+template <typename Type>
+void set_values_union(Type& value, UnionDisc disc)
+{
+  set_base_values_union(value, disc);
+}
+
+template<typename TypeA, typename TypeB>
+void amalgam_serializer_test_union(const Encoding& encoding, const DataView& expected_cdr, UnionDisc disc)
+{
+  TypeA value;
+  set_values_union(value, disc);
+  TypeB result;
+  amalgam_serializer_test<TypeA, TypeB>(encoding, expected_cdr, value, result);
+}
+
+template<typename Type>
+void serializer_test_union(const Encoding& encoding, const DataView& expected_cdr, UnionDisc disc)
+{
+  amalgam_serializer_test_union<Type, Type>(encoding, expected_cdr, disc);
+}
+
+template<typename Type>
+void baseline_checks_union(const Encoding& encoding, const DataView& expected_cdr, UnionDisc disc)
+{
+  Type value;
+  value._d(disc);
+
+  EXPECT_EQ(serialized_size(encoding, value), expected_cdr.size);
+
+  serializer_test_union<Type>(encoding, expected_cdr, disc);
+}
```

Comment:
You should be able to eliminate these three functions by defining overloads for `set _values` and `expect_values_equal` for the unions.

Reply:
Agreed to keep this logic, given that union set_value function should take additional 'discriminator' parameter.